src/Storages/MergeTree/MergeTreeSource.cpp

```diff
@@ -0,0 +1,220 @@
+#include <Storages/MergeTree/MergeTreeSource.h>
+#include <Storages/MergeTree/MergeTreeBaseSelectProcessor.h>
+#include <Interpreters/threadPoolCallbackRunner.h>
+#include <IO/IOThreadPool.h>
+#include <Common/EventFD.h>
+
+namespace DB
+{
+
+MergeTreeSource::MergeTreeSource(MergeTreeSelectAlgorithmPtr algorithm_)
+    : ISource(algorithm_->getHeader())
+    , algorithm(std::move(algorithm_))
+{
+#if defined(OS_LINUX)
+    if (algorithm->getSettings().use_asynchronous_read_from_pool)
+        async_reading_state = std::make_unique<AsyncReadingState>();
+#endif
+}
+
+MergeTreeSource::~MergeTreeSource() = default;
+
+std::string MergeTreeSource::getName() const
+{
+    return algorithm->getName();
+}
+
+void MergeTreeSource::onCancel()
+{
+    algorithm->cancel();
+}
+
+#if defined(OS_LINUX)
+struct MergeTreeSource::AsyncReadingState
+{
+    /// NotStarted -> InProgress -> IsFinished -> NotStarted ...
```

Comment:
```suggestion
    /// NotStarted -> InProgress -> IsFinished -> NotStarted
```

if I'm not missing smth, there cannot be cycle. in other words state can be `InProgress` only once in a lifetime.

Reply:
No, we don't recreae AsyncReadingState.
It's created once in ctor and a cycle `NotStarted -> InProgress -> IsFinished` is done for every generated chunk.